include/deal.II/dofs/dof_vector.h

```diff
@@ -0,0 +1,243 @@
+#include <deal.II/lac/constraint_matrix.h>
+#include <deal.II/lac/vector.h>
+#include <deal.II/lac/vector_memory.h>
+
+DEAL_II_NAMESPACE_OPEN
+/**
+ * An object storing a vector describing a finite element function
+ * together with its DoFHandler and the ConstraintMatrix associated
+ * with the handler.
+ *
+ * Data vectors for finite element discretizations do not make sense
+ * without their DoFHandler. Nevertheless, deal.II has assumed for
+ * decades, that there is a single DoFHandler in a program and that
+ * all vectors refer to it. This class encapsulates both, to begin
+ * with in a minimally intrusive way, giving access to both
+ * structures.
+ *
+ * @todo Ensure that both objects are synchronized, that is, the
+ * DoFHandler is not changes without the vector being resized.
+ */
+template <class DH, class VECTOR=Vector<double> >
+class DoFVector :  public Subscriptor
+{
+public:
+  /// @name Constructors and initialization
+  ///@{
+  /**
+   * Construct an object with a DoFHandler and borrow a vector from
+   * GrowingVectorMemory for its data. This vector is then owned by
+   * this object and released when the object itself expires.
+   */
+  DoFVector (const DH &dh, const ConstraintMatrix &constraints);
+
+  /**
+   * Construct an object with a DoFHandler but without constraints and
+   * borrow a vector from GrowingVectorMemory for its data. This
+   * vector is then owned by this object and released when the object
+   * itself expires.
+   *
+   * @note This constructor uses the ConstraintMatrix
+   * #no_constraints. The created object thus can output vectors in
+   * distributed format, but writing into vectors will most likely
+   * fail without using a constraint matrix from somewhere else.
+   */
+  DoFVector (const DH &dh);
+
+  /**
+   * Construct an object with a DoFHandler and borrow a vector from
+   * GrowingVectorMemory for its data. This vector is not owned by the
+   * object and read-access only.
+   *
+   * @note An object created this way is effectively constant. In
+   * order to avoid obscure access errors to the data vector, it
+   * should also be declared `const`.
+   */
+  DoFVector (const DH &dh, const ConstraintMatrix &constraints, const VECTOR &v);
+
+  /**
+   * Construct an object with a DoFHandler but without constraints and
+   * borrow a vector from GrowingVectorMemory for its data. This
+   * vector is not owned by the object and read-access only.
+   *
+   * @note An object created this way is effectively constant. In
+   * order to avoid access errors to the data vector, it should also
+   * be declared `const`. Also, vectors are expected to be in
+   * distributed format, since the information on constraints is
+   * missing.
+   */
+  DoFVector (const DH &dh, const VECTOR &v);
+
+  /**
+   * Destructor, deleting own data.
+   */
+  ~DoFVector ();
+
+  /**
+   * Modify the vector dimensions to match the current state of the
+   * dof handler.
+   */
+  void sync ();
```

Comment:
what happens to the content of the vector? Is it just reset to zero? Is it transferred?